src/operations/shared.ts

```diff
@@ -0,0 +1,62 @@
+import { FieldNode } from 'graphql';
+import { Fragments, Variables, SelectionSet } from '../types';
+import { Store } from '../store';
+import { joinKeys, keyOfField } from '../helpers';
+
+import {
+  getTypeCondition,
+  getFieldArguments,
+  shouldInclude,
+  isFieldNode,
+  isInlineFragment,
+  getSelectionSet,
+  getName,
+} from '../ast';
+
+interface Context {
+  store: Store;
+  variables: Variables;
+  fragments: Fragments;
+}
+
+export const forEachFieldNode = (
+  typeName: string,
+  entityKey: string,
+  select: SelectionSet,
+  ctx: Context,
+  cb: (node: FieldNode) => void
+) => {
+  select.forEach(node => {
+    if (!shouldInclude(node, ctx.variables)) {
+      // Directives instruct this node to be skipped
+      return;
+    } else if (!isFieldNode(node)) {
+      // A fragment is either referred to by FragmentSpread or inline
+      const def = isInlineFragment(node) ? node : ctx.fragments[getName(node)];
+
+      if (def !== undefined) {
+        const fragmentSelect = getSelectionSet(def);
+        const typeCondition = getTypeCondition(def);
+
+        const matches =
+          typeCondition === typeName ||
+          !fragmentSelect.some(node => {
```

Comment:
If we quit on first occurrence we can't notify the consumer of the graphcache what exactly was wrong about the fragment. I'd prefer this for DX purposes

Reply:
I wouldn't actually notify though unless we have a solution for non-heuristic fragment matching. If the user uses interfaces then they'll be aware that they are, but we can't offer them any advice on when a fragment matches or doesn't match 😓